core/libp2p/event/bus.hpp

```diff
@@ -0,0 +1,214 @@
+      return connections > 0;
+    }
+
+   private:
+    Channel() = default;
+    virtual ~Channel() = default;
+
+    /**
+     * Proper deleter for type-erased channel
+     * @note no type checking is performed at this level
+     * @param erased_channel_ptr
+     */
+    static void deleter(void *erased_channel_ptr) {
+      auto ptr = reinterpret_cast<Channel *>(erased_channel_ptr);
+      delete ptr;
+    }
+
+    /**
+     * Get the channel back from an erased pointer
+     * @param ptr - the type-erased channel pointer
+     * @return - the type safe channel pointer
+     */
+    static Channel *get_channel(erased_channel_ptr &ptr) {
+      return reinterpret_cast<Channel *>(ptr.get());
+    }
+
+    /**
+     * Construct a unique_ptr for the type erased method poiner
+     */
+    static erased_channel_ptr make_unique() {
+      return erased_channel_ptr(new Channel(), &deleter);
+    }
+
+    boost::signals2::signal<void(const Data &), DispatchPolicy> signal_;
+
+    friend class Bus;
+  };
+
+  /**
+   * Declaration of the channel, which is to be used in Bus
+   * @tparam Tag - API specific discriminator used to distinguish between
+   * otherwise identical data types
+   * @tparam Data - the type of the Data the channel carries
+   * @tparam DispatchPolicy - The dispatch policy to use for this channel
+   * (defaults to @ref drop_exceptions)
+   */
+  template <typename Tag, typename Data,
+            typename DispatchPolicy = drop_exceptions>
+  struct channel_decl {
+    using channel_type = Channel<Data, DispatchPolicy>;
+    using tag_type = Tag;
+  };
+
+  template <typename... Ts>
+  std::true_type is_channel_decl_impl(const channel_decl<Ts...> *);
+
+  std::false_type is_channel_decl_impl(...);
+
+  template <typename T>
+  using is_channel_decl = decltype(is_channel_decl_impl(std::declval<T *>()));
+
+  /**
+   * Event bus, containing channels and providing a convenient access to them
+   */
+  class Bus {
+   public:
+    /**
+     * Fetch a reference to the channel declared by the passed in type. This
+     * will construct the channel on first access
+     * @tparam ChannelDecl - @ref libp2p::event::channel_decl
+     * @return reference to the channel described by the declaration
+     */
+    template <typename ChannelDecl>
+    auto getChannel()
+        -> std::enable_if_t<is_channel_decl<ChannelDecl>::value,
+                            typename ChannelDecl::channel_type &> {
+      using channel_type = typename ChannelDecl::channel_type;
+      auto key = std::type_index(typeid(ChannelDecl));
+      auto itr = channels_.find(key);
+      if (itr != channels_.end()) {
+        return *channel_type::get_channel(itr->second);
+      } else {
+        channels_.emplace(std::make_pair(key, channel_type::make_unique()));
+        return *channel_type::get_channel(channels_.at(key));
+      }
+    }
+
+   private:
+    using erased_channel_ptr = std::unique_ptr<void, void (*)(void *)>;
```

Comment:
you can remove this, since it is defined here https://github.com/soramitsu/kagome/pull/158/files#diff-ebfe0d83f5a1ecc31ccf898bc9b22fa4R22